### PR TITLE
Add Read return values function

### PR DESCRIPTION
**Pull request by mmsc2**

### CODE REVIEW

File: src/vm/errors/runner_errors.rs [hunk 1]
```diff
@@ -73,7 +73,7 @@ pub enum RunnerError {
     BuiltinSegmentNameCollision(&'static str),
     #[error("Error while finalizing segments: {0}")]
     FinalizeSegements(MemoryError),
-    #[error("finalize_segments called but proof_mode is not enabled")]
+    #[error("Finalize_segments called but proof_mode is not enabled")]
```
Comment (Oppen) [inline src/vm/errors/runner_errors.rs:76]:
```suggestion
    #[error("finalize_segments called but proof_mode is not enabled")]
```

The function name is lower case.

File: src/vm/runners/cairo_runner.rs [hunk 2]
```diff
@@ -1022,6 +1022,44 @@ impl CairoRunner {
...
+    pub fn read_return_values(&mut self, vm: &VirtualMachine) -> Result<(), RunnerError> {
+        if !self.run_ended {
+            return Err(RunnerError::FinalizeNoEndRun);
+        }
+        let mut pointer = vm.get_ap();
+        for builtin_name in self.program.builtins.iter().rev() {
+            let builtin_runner = vm
+                .builtin_runners
+                .iter()
+                .find(|(name, _builtin)| builtin_name == name);
```
Comment (Oppen) [inline src/vm/runners/cairo_runner.rs:1035]: I think this is a `HashMap`, so just do `builtin_runners.get(name)`.

File: src/vm/runners/cairo_runner.rs [hunk 3]
```diff
@@ -1022,6 +1022,44 @@ impl CairoRunner {
...
+                Some((_, builtin)) => {
+                    let (new_pointer, _) = builtin.final_stack(vm, pointer)?;
+                    pointer = new_pointer;
+                }
+            }
+        }
+        if !self.segments_finalized {
+            return Err(RunnerError::FailedAddingReturnValues);
+        }
+        // use extend self.public_memory
```
Comment (Oppen) [inline src/vm/runners/cairo_runner.rs:1048]: Lingering comment.

File: src/vm/runners/cairo_runner.rs [hunk 4]
```diff
@@ -3458,6 +3496,28 @@ mod tests {
...
+        assert!(cairo_runner.segments_finalized);
+        //Check values written by first call to segments.finalize()
+        assert_eq!(vm.segments.segment_sizes.get(&0), Some(&8_usize));
+
+        assert_eq!(cairo_runner.read_return_values(&vm), Ok(()));
+        assert_eq!(
+            cairo_runner.execution_public_memory.unwrap(),
+            Vec::<usize>::new()
+        );
+    }
```
Comment (Oppen) [inline src/vm/runners/cairo_runner.rs:3519]: We also need to test error cases and cases where there's public memory after running.

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/runners/cairo_runner.rs [hunk 5]
```diff
@@ -3458,6 +3496,28 @@ mod tests {
...
+        cairo_runner.run_ended = true;
+        let mut vm = vm!();
+        assert_eq!(cairo_runner.finalize_segments(&mut vm), Ok(()));
+        assert!(cairo_runner.segments_finalized);
+        //Check values written by first call to segments.finalize()
+        assert_eq!(vm.segments.segment_sizes.get(&0), Some(&8_usize));
+
+        assert_eq!(cairo_runner.read_return_values(&vm), Ok(()));
+        assert_eq!(
+            cairo_runner.execution_public_memory.unwrap(),
```
Comment (Juan-M-V) [inline src/vm/runners/cairo_runner.rs:3516]: Please use expect instead of unwrap!

File: src/vm/runners/cairo_runner.rs (hunk 2 above)
Comment (mmsc2) [reply src/vm/runners/cairo_runner.rs:1035]: builtin_runners is actually a vec<(string, builtin_runner)>, that is why an iterator is used

File: src/vm/errors/runner_errors.rs (hunk 1 above)
Comment (mmsc2) [reply src/vm/errors/runner_errors.rs:76]: fixed

File: src/vm/runners/cairo_runner.rs (hunk 5 above)
Comment (mmsc2) [reply src/vm/runners/cairo_runner.rs]: changed

**Review by fmoletta — APPROVED** [review]

**Review by Oppen — APPROVED** [review]

**Review by juanbono — APPROVED** [review]
LGTM

**Pull request merged**